src/server/scripts/Spells/spell_druid.cpp

```diff
@@ -811,23 +820,78 @@ class spell_dru_item_t6_trinket : public AuraScript
 };
 
 // 33763 - Lifebloom
+// 188550 - Lifebloom (Undergrowth)
 class spell_dru_lifebloom : public AuraScript
 {
-    bool Validate(SpellInfo const* /*spell*/) override
+    bool Validate(SpellInfo const* /*spellInfo*/) override
     {
         return ValidateSpellInfo({ SPELL_DRUID_LIFEBLOOM_FINAL_HEAL });
     }
 
-    void AfterRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    void HandleAfterRemove(AuraEffect const* aurEff, AuraEffectHandleModes /*mode*/)
     {
-        // Final heal only on duration end
+        Unit* caster = GetCaster();
+        Unit* target = GetTarget();
+        if (!caster)
+            return;
+
         if (GetTargetApplication()->GetRemoveMode() == AURA_REMOVE_BY_EXPIRE || GetTargetApplication()->GetRemoveMode() == AURA_REMOVE_BY_ENEMY_SPELL)
-            GetCaster()->CastSpell(GetUnitOwner(), SPELL_DRUID_LIFEBLOOM_FINAL_HEAL, true);
+            caster->CastSpell(target, SPELL_DRUID_LIFEBLOOM_FINAL_HEAL, CastSpellExtraArgs(aurEff));
+    }
+
+    void Register() override
+    {
+        AfterEffectRemove += AuraEffectRemoveFn(spell_dru_lifebloom::HandleAfterRemove, EFFECT_0, SPELL_AURA_PERIODIC_HEAL, AURA_EFFECT_HANDLE_REAL);
+    }
+};
+
+// 33778 - Lifebloom (Heal)
+class spell_dru_lifebloom_heal : public SpellScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({ SPELL_DRUID_EFFLORESCENCE })
+            && ValidateSpellEffect({ { SPELL_DRUID_VERDANCY, EFFECT_0 } });
+    }
+
+    void HandleOnHit(SpellEffIndex /*effIndex*/)
+    {
+        Unit* caster = GetCaster();
+
+        // Note: Verdancy talent.
+        if (AuraEffect* const verdancyEffect = caster->GetAuraEffect(SPELL_DRUID_VERDANCY, EFFECT_0))
+        {
+            AreaTrigger* const efflorescenceAT = caster->GetAreaTrigger(SPELL_DRUID_EFFLORESCENCE);
+            if (!efflorescenceAT)
+                return;
+
+            std::list<Unit*> targetList;
+
+            for (ObjectGuid const& guid : efflorescenceAT->GetInsideUnits())
+            {
+                if (Unit* unit = ObjectAccessor::GetUnit(*caster, guid))
+                {
+                    if (!caster->IsValidAssistTarget(unit))
+                        continue;
+
+                    targetList.push_back(unit);
+                }
+            }
+
+            // Note: max. targets is Verdancy's EFFECT_0.
+            uint32 const maxTargets = verdancyEffect->GetAmount();
+
+            if (targetList.size() > maxTargets)
```

Comment:
You could random shuffle efflorescenceAT->GetInsideUnits() into a new list, iterate over it and then get the unit, check if its valid and cast on it in one go.

You basically skip 1 for loop with that, but you need to count if you have your max targets then.